payments/payments.go

```diff
@@ -0,0 +1,41 @@
+package payments
+
+import (
+	"context"
+	"net/http"
+)
+
+const (
+	StripeProvider = "stripe"
+	PayPalProvider = "paypal"
+)
+
+// Provider represents a payment provider that can optionally charge, refund,
+// preauthorize payments.
+type Provider interface {
```

Comment:
I think that you're actually not creating anything yet. You haven't written to the DB. It is all inside a transaction. All we did was validate you haven't lied yet. 

You do raise a good point though. What about instead of making each of those an interface they were just functions that were typed? 

```
type Charger func(string, string)

type Provider interface {
  NewCharger() Charger
}

type implProvider struct {
  
}

func (ip implProvider) NewCharger() Charger {
  return ip.charge
}
```